File: typescript/src/phone/phoneClient.ts
import { EDVSError } from "../exceptions/edvsException";
import { getPhoneValidateRequestFromConfig } from "../server/phone/restApiPhoneValidateRequest";
import { RestApiStubImpl } from "../server/restApiStub";
import { PhoneConfiguration } from "./phoneConfiguration";
import { PhoneValidateResult, restApiResponseToPhoneValidateResult } from "./validate/phoneValidateResult";

/**
 * Client class for interacting with the phone-related APIs.
 * Provides methods for validating phone numbers.
 */
export class PhoneClient {

    private readonly configuration: PhoneConfiguration;
    private readonly restApiStub: RestApiStubImpl;

    /**
     * Initializes a new instance of the {@link PhoneClient} class with the specified configuration.
     *
     * @param configuration The configuration settings for the phone client.
     */
    public constructor(configuration: PhoneConfiguration) {
        this.configuration = configuration;
        this.restApiStub = new RestApiStubImpl(configuration);
    }

    /**
     * Validates a phone number asynchronously.
     *
     * @param phoneNumber The phone number to validate.
     * @return A promise that resolves to the validation result.
     * @throws EDVSError If the API response contains an error.
     */
    public async validate(phoneNumber: string): Promise<PhoneValidateResult> {
        const headers = this.configuration.getCommonHeaders(false);
        const request = getPhoneValidateRequestFromConfig(this.configuration);
        request.number = phoneNumber;

        if (this.configuration.options.includeMetadata) {
            headers.set("Add-Metadata", "true" as String); // NOSONAR
        }
        try {
            const resp = await this.restApiStub.validatePhoneV2(request, headers);
            if (resp.error) {
                throw EDVSError.using(resp.error);
            }
            return restApiResponseToPhoneValidateResult(resp);
        } catch (error) {
            return Promise.reject(error instanceof Error ? error : new Error(String(error)));
        }
    }
}

File: java/src/main/java/com/experian/dvs/client/phone/PhoneClient.java
package com.experian.dvs.client.phone;

import com.experian.dvs.client.exceptions.RestApiInterruptionOrExecutionException;
import com.experian.dvs.client.phone.validate.PhoneResult;
import com.experian.dvs.client.phone.validate.PhoneResultFuture;
import com.experian.dvs.client.server.RestApiAsyncImpl;
import com.experian.dvs.client.server.RestApiAsyncStub;
import com.experian.dvs.client.server.phone.RestApiPhoneValidateRequest;
import com.experian.dvs.client.server.phone.RestApiPhoneValidateResponse;

import java.io.Closeable;
import java.io.IOException;
import java.util.Map;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;

/**
 * AddressClient class for interacting with the phone-related APIs.
 * Provides methods for validating phone numbers.
 */
public class PhoneClient implements Closeable {

    private final PhoneConfiguration configuration;
    private final RestApiAsyncStub restApiAsyncStub;

    /**
     * Initializes a new instance of the {@link PhoneClient} class with the specified configuration.
     *
     * @param configuration The configuration settings for the client.
     */
    public PhoneClient(PhoneConfiguration configuration) {
        this.configuration = configuration;
        this.restApiAsyncStub = new RestApiAsyncImpl(configuration);
    }

    /**
     * Closes the client and releases any resources held by it.
     *
     * @throws IOException If an I/O error occurs while closing the client.
     */
    @Override
    public void close() throws IOException {
        this.restApiAsyncStub.close();
    }

    /**
     * Validates an phone number synchronously.
     *
     * @param phoneNumber The email address to validate.
     * @return The validation result.
     * @throws RestApiInterruptionOrExecutionException If the operation is interrupted or fails.
     */
    public PhoneResult validate(final String phoneNumber) {
        return validate(phoneNumber, "");
    }

    /**
     * Validates an email address synchronously.
     *
     * @param phoneNumber The email address to validate.
     * @param referenceId The reference ID for tracking the request.
     * @return The validation result.
     * @throws RestApiInterruptionOrExecutionException If the operation is interrupted or fails.
     */
    public PhoneResult validate(final String phoneNumber, final String referenceId) {
        try {
            return this.validateAsync(phoneNumber, referenceId).get();
        } catch (InterruptedException | ExecutionException e) {
            throw new RestApiInterruptionOrExecutionException(e);
        }
    }

    /**
     * Validates an email address asynchronously.
     *
     * @param phoneNumber The email address to validate.
     * @return A future representing the asynchronous operation.
     */
    public Future<PhoneResult> validateAsync(final String phoneNumber) {
        return validateImpl(phoneNumber, "");
    }

    /**
     * Validates an email address asynchronously.
     *
     * @param phoneNumber The email address to validate.
     * @param referenceId The reference ID for tracking the request.
     * @return A future representing the asynchronous operation.
     */
    public Future<PhoneResult> validateAsync(final String phoneNumber, final String referenceId) {
        return validateImpl(phoneNumber, referenceId);
    }

    private Future<PhoneResult> validateImpl(final String phoneNumber, final String referenceId) {
        RestApiPhoneValidateRequest request = RestApiPhoneValidateRequest.using(configuration);
        request.setNumber(phoneNumber);
        final Map<String, Object> headers = this.configuration.getCommonHeaders(referenceId);

        if (this.configuration.getMetadata()) {
            headers.put("Add-Metadata", Boolean.TRUE.toString());
        }

        final Future<RestApiPhoneValidateResponse> response = this.restApiAsyncStub.validatePhoneV2(request, headers);
        return new PhoneResultFuture(response);
    }
}

